blockchain/internal/spendpruner/pruner.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC
+// license that can be found in the LICENSE file.
+
+package spendpruner
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/decred/dcrd/blockchain/v4/indexers"
+	"github.com/decred/dcrd/chaincfg/chainhash"
+)
+
+// SpendJournalPruner represents a spend journal data pruner that ensures
+// data needed by spend consumers are retained until no longer needed.
+type SpendJournalPruner struct {
+	queryer    indexers.ChainQueryer
+	dependents map[chainhash.Hash][]string
+	consumers  map[string]SpendConsumer
+	pruneCh    chan *chainhash.Hash
+	mtx        sync.RWMutex
+}
+
+// Ensure SpendJournalPruner implements the SpendJPruner interface.
+var _ SpendPruner = (*SpendJournalPruner)(nil)
+
+// newSpendJournalPruner initializes a spend journal pruner.
+func newSpendJournalPruner(ctx context.Context, queryer indexers.ChainQueryer) *SpendJournalPruner {
+	prunefilter := &SpendJournalPruner{
+		queryer:    queryer,
+		dependents: make(map[chainhash.Hash][]string),
+		consumers:  make(map[string]SpendConsumer),
+		pruneCh:    make(chan *chainhash.Hash),
+	}
+
+	go prunefilter.handleSpendPrunes(ctx)
+
+	return prunefilter
+}
+
+// UpdateConsumerTip updates the tip hash of the associated spend journal
+// consumer with the provided id.
+func (s *SpendJournalPruner) UpdateConsumerTip(id string, tipHash *chainhash.Hash) error {
```

Comment:
Where/when would `UpdateConsumerTip` be called, and how does this fit into the overall picture?  It seems like it would be up to the consumers to manage this and not the `SpendJournalPruner` since the `SpendJournalPruner` is responsible for removing spend journal entries but not keeping consumers in sync.

Reply:
`UpdateConsumerTip` will be called when the represented consumer's tip hash is updated. Considering the pruner keeps a list of consumers I think it's best it provides an update function for the represented consumers to sync their updated state.

Reply:
I might be missing something else, but I still am not following why the `SpendJournalPruner` would be updating the tip on consumers.  The `SpendJournalPruner` is just managing the spend journal removal, not keeping the consumers in sync.  Further, it is not even tracking the tip or the chain itself, it is just tracking spend journal entries that are pending removal.

Reply:
The `SpendJournalPruner` will not b updating the tip of consumers. It holds references of those consumers and `UpdateConsumerTip` provides access to updating those references. `UpdateConsumerTip` will be called by the consumers, for example for the indexers, when a new block is connected `UpdateConsumerTip` will be called to sync the state of the consumer with the reference the pruner has.

Reply:
> UpdateConsumerTip will be called by the consumers, for example for the indexers, when a new block is connected UpdateConsumerTip will be called to sync the state of the consumer with the reference the pruner has.

I'm not clear on why this would be the responsibility of the `SpendJournalPruner`.  Is the `SpendJournalPruner` only intended to prune spend journal entries, or is it also syncing the state of consumers?  These seem like two separate things, and it's hard to tell how this function fits in since it's only used in the test.

Reply:
Will start integrating this into the async indexers work, think that'd best show my intentions for `UpdateConsumerTip` or further refine the `SpendConsumer` interface.